File: leetcode/java/0001. Two Sum.java
package leetcode.java;

import java.util.HashMap;

// Time complexity: O(N)
// Space complexity: O(N)

class Solution {
    public static int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> map = new HashMap<>();
        for (int i=0; i<nums.length; i++) {
            int remainder = target - nums[i];
            if (map.containsKey(remainder)) {
                return new int[]{i, map.get(remainder)};
            } else {
                map.put(nums[i], i);
            }
        }
        return new int[]{-1, -1};
    }
}

File: leetcode/python/0001. Two Sum.py
# Time complexity: O(N)
# Space complexity: O(N)

class Solution(object):
    def twoSum(self, nums, target):
        storage = {}
        
        for idx, num in enumerate(nums):
            remainder = target - num
            if remainder in storage:
                return [idx, storage[remainder]]
            else:
                storage[num] = idx
        return []

instance = Solution()
print(instance.twoSum([2,5,7,9], 12))